packages/lit-html/src/directives/cache.ts

```diff
@@ -57,38 +57,42 @@ export const cache = directive(
     }
 
     update(containerPart: ChildPart, [v]: DirectiveParameters<this>) {
-      // If the new value is not a TemplateResult from the same Template as the
-      // previous value, move the nodes from the DOM into the cache.
+      // If the previous value is a TemplateResult and the new value is not,
+      // or is a different Template as the previous value, move the child part
+      // from into the cache.
       if (
-        this.value !== undefined &&
-        this.value.strings !== (v as TemplateResult).strings
-      ) {
+        isTemplateResult(this.value) && 
+        (!isTemplateResult(v) || this.value.strings !== v.strings)) {
         // This is always an array because we return [v] in render()
         const partValue = getComittedValue(containerPart) as Array<ChildPart>;
         const childPart = partValue.pop()!;
         let cachedContainerPart = this.templateCache.get(this.value.strings);
         if (cachedContainerPart === undefined) {
           const fragment = new DocumentFragment();
           cachedContainerPart = render(nothing, fragment);
-          setComittedValue(cachedContainerPart, [childPart]);
           this.templateCache.set(this.value.strings, cachedContainerPart);
         }
         // Move into cache
-        insertPart(containerPart, undefined, childPart);
+        setComittedValue(cachedContainerPart, [childPart]);
+        insertPart(cachedContainerPart, undefined, childPart);
         clearPart(containerPart);
```

Comment:
Should the clear here be necessary if we've definitely just moved the whole `childPart` which is what would have been committed there into the other part?

Reply:
ah, good point